odoo/tools/xml_utils.py

```diff
@@ -1,22 +1,38 @@
 # -*- coding: utf-8 -*-
+import base64
+from io import BytesIO
 from lxml import etree
-from odoo.tools.misc import file_open
+
 from odoo.exceptions import UserError
 
 def check_with_xsd(tree_or_str, stream):
     raise UserError("Method 'check_with_xsd' deprecated ")
 
+class odoo_resolver(etree.Resolver):
+    def __init__(self, env):
+        self.env = env
+
+    def resolve(self, url, id, context):
+        attachment = self.env['ir.attachment'].search([('name', '=', url)])
+        if attachment:
+            return self.resolve_string(base64.b64decode(attachment.datas), context)
+
 
-def _check_with_xsd(tree_or_str, stream):
+def _check_with_xsd(tree_or_str, stream, env=None):
+    parser = etree.XMLParser()
     if not isinstance(tree_or_str, etree._Element):
         tree_or_str = etree.fromstring(tree_or_str)
-    xml_schema_doc = etree.parse(stream)
+    # Enable searching in the ir.attachment's if env is given.
+    if env:
+        parser.resolvers.add(odoo_resolver(env))
+        if isinstance(stream, str) and stream.endswith('.xsd'):
+            attachment = env['ir.attachment'].search([('name', '=', stream)])
```

Comment:
We should give an error when the file is not found.  But maybe something specific to catch.  (FileNotFoundError??)  I know it is in most cases not dramatic when the xsd is not checked, but it can be in handy to have it at least in the logs if the xsds were not downloaded.  (actually for me, these three lines can also be removed, although it renders a lot of code shorter)

Reply:
I added these lines because a lot of code was indeed duplicated elsewhere.

If I'm not mistaken, building the schema will fail if no resolver can find the file; so an error should be raised by `lxml` at that point.
I don't think a resolver should fail if it doesn't find something because other resolvers might.

Reply:
I changed the test a little bit to show a failing XSD parser, is it clearer?